### PR TITLE
Add support for async sleep functions in tenacity.retry annotation.

### PR DESCRIPTION
**Pull request by jakkdl**
Fixes https://github.com/jd/tenacity/issues/399

Since tests aren't typechecked I didn't add a test for this, although if you want I could open another PR with tox modifications and `# mypy: disable-error-codes= I added to check it locally.

### CODE REVIEW

File: tenacity/__init__.py [hunk 1]
```diff
@@ -499,7 +499,7 @@ def retry(func: WrappedFn) -> WrappedFn:
 
 @t.overload
 def retry(
-    sleep: t.Callable[[t.Union[int, float]], None] = sleep,
+    sleep: t.Callable[[t.Union[int, float]], t.Union[t.Awaitable[None], None]] = sleep,
```
Comment (jd) [inline tenacity/__init__.py:502]:
```suggestion
    sleep: t.Callable[[t.Union[int, float]], t.Optional[t.Awaitable[None]] = sleep,
```

**Comment by jd** [comment]
We have mypy running but it's possible that's missing some tests, if you have some that'd be great.

File: tenacity/__init__.py (hunk 1 above)
Comment (jakkdl) [reply tenacity/__init__.py]: oh yeah of course, I've gotten too used to `from __future__ import annotations` and just doing `| None`

**Comment by Zac-HD** [comment]
@jd any timeline on this, or things I can do to help?  It's currently pretty painful to use tenacity in async codebases and I'd love to get the patch shipped asap 🙏

**Comment by jakkdl** [comment]
@jd any blocker for going ahead and merging this?

**Comment by jakkdl** [comment]
Thanks!

**Pull request merged**